### PR TITLE
Fix logging caller for helper functions and expose configuration options

### PR DESCRIPTION
**Pull request by alfrunes**

### CODE REVIEW

File: log/log.go [hunk 1]
```diff
@@ -59,17 +64,92 @@ type ContextLogger interface {
 
 // init initializes the global logger to sane defaults.
 func init() {
-	Log.Formatter = &logrus.TextFormatter{
-		FullTimestamp: true,
+	switch strings.ToLower(os.Getenv("LOG_FORMAT")) {
+	case "ndjson", "json":
```
Comment (merlin-northern) [inline log/log.go:68]: could we have consts for these?

File: log/log.go [hunk 2]
```diff
@@ -121,27 +201,49 @@ func (hook ContextHook) Levels() []logrus.Level {
...
+	var pcs [3]uintptr
+	if _, ok := entry.Data["caller"]; !ok {
+		// We don't know how deep we are in the callstack since the hook can be fired
+		// at different levels. Search between depth 6 -> 8.
+		i := runtime.Callers(6, pcs[:])
+		frames := runtime.CallersFrames(pcs[:i])
+		var caller *runtime.Frame
+		for frame, ok := frames.Next(); ok; frame, ok = frames.Next() {
+			if !strings.HasPrefix(frame.Function, "github.com/sirupsen/logrus") {
+				caller = &frame
```
Comment (merlin-northern) [inline log/log.go:223]: I do not like the magic numbers, and it all seems a bit cryptic and a bit messy. why are we doing it in this way?

**Review by merlin-northern — COMMENTED** [review]
this is my first time in this file at this level, it seems all a bit complicated.

File: log/log.go (hunk 1 above)
Comment (alfrunes) [reply log/log.go:68]: This was just a simple hack to make it easier for me to configure it on Hosted Mender. I didn't mean for it to become a feature.

File: log/log.go (hunk 2 above)
Comment (alfrunes) [reply log/log.go]: Added more comments explaining this in the source code.

File: log/log.go [hunk 3]
```diff
@@ -121,27 +210,57 @@ func (hook ContextHook) Levels() []logrus.Level {
...
-	//for Errorf(), Warnf(), etc. - we have to skip 1 lvl up
-	for skip := 6; skip < 8; skip++ {
-		if pc, file, line, ok := runtime.Caller(skip); ok {
-			funcName := runtime.FuncForPC(pc).Name()
-
-			//detect if we're still in logrus (formatting funcs)
-			if !strings.Contains(funcName, "github.com/sirupsen/logrus") {
-				entry.Data["file"] = path.Base(file)
-				entry.Data["func"] = path.Base(funcName)
-				entry.Data["line"] = line
```
Comment (merlin-northern) [inline log/log.go:136]: ok you have to educate me a bit here.  we used those, as they are well-defined consts from logrus (e.g.:FieldKeyFunc). now we are only using "caller" which is not a well defined const and is unused in logrus code. how does it work?

File: log/log.go [hunk 4]
```diff
@@ -121,27 +210,57 @@ func (hook ContextHook) Levels() []logrus.Level {
...
+	var pcs [1 + maxCallDepth - minCallDepth]uintptr
+	if _, ok := entry.Data["caller"]; !ok {
+		// We don't know how deep we are in the callstack since the hook can be fired
+		// at different levels. Search between depth 6 -> 8.
+		i := runtime.Callers(minCallDepth, pcs[:])
+		frames := runtime.CallersFrames(pcs[:i])
+		var caller *runtime.Frame
+		for frame, ok := frames.Next(); ok; frame, ok = frames.Next() {
+			if !strings.HasPrefix(frame.Function, "github.com/sirupsen/logrus") {
+				caller = &frame
```
Comment (merlin-northern) [inline log/log.go:236]: I know I am annoying, but could we please define `"caller"` and `"github.com/sirupsen/logrus"`?

**Review by merlin-northern — COMMENTED** [review]
I actually asked for education in this review. looks good, just make me calm.

File: log/log.go (hunk 3 above)
Comment (alfrunes) [reply log/log.go:136]: I wasn't aware that these were "well-defined logrus consts". It just seems pointless to index the line number, filename and function independently in the log parser (or indexing the caller context at all seems pretty pointless), so I condensed it to a single field.

File: log/log.go [hunk 5]
```diff
@@ -44,6 +49,15 @@ var (
 	Log = logrus.New()
 )
 
+const (
+	envLogFormat        = "LOG_FORMAT"
+	envLogLevel         = "LOG_LEVEL"
+	envLogDisableCaller = "LOG_DISABLE_CALLER"
```
[the remark this answers is not in the document]
Comment (alfrunes) [reply log/log.go:55]: Any suggestions?

[the remark this answers is not in the document]
Comment (alfrunes) [reply log/log.go]: Changed it to `LOG_DISABLE_CALLER_CONTEXT`

**Review by merlin-northern — CHANGES_REQUESTED** [review]
```diff
diff --git a/log/log.go b/log/log.go
index 80e1baa..c9823c7 100644
--- a/log/log.go
+++ b/log/log.go
@@ -39,7 +39,6 @@ import (
        "runtime"
        "strconv"
        "strings"
-       "time"

        "github.com/sirupsen/logrus"
 )
```

what I did is I tried with tenantadm and this patch: I have 0 logs printed.

**Comment by alfrunes** [comment]
> ```diff
> diff --git a/log/log.go b/log/log.go
> index 80e1baa..c9823c7 100644
> --- a/log/log.go
> +++ b/log/log.go
> @@ -39,7 +39,6 @@ import (
>         "runtime"
>         "strconv"
>         "strings"
> -       "time"
> 
>         "github.com/sirupsen/logrus"
>  )
> ```
> 
> what I did is I tried with tenantadm and your patch: I have 0 logs printed.

I made a couple of mistakes on my last force-push. They should be corrected now.

**Comment by merlin-northern** [comment]
> > ```diff
> > diff --git a/log/log.go b/log/log.go
> > index 80e1baa..c9823c7 100644
> > --- a/log/log.go
> > +++ b/log/log.go
> > @@ -39,7 +39,6 @@ import (
> >         "runtime"
> >         "strconv"
> >         "strings"
> > -       "time"
> > 
> >         "github.com/sirupsen/logrus"
> >  )
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > what I did is I tried with tenantadm and your patch: I have 0 logs printed.
> 
> I made a couple of mistakes on my last force-push. They should be corrected now.

```bash
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x1436b36]

goroutine 1 [running]:
github.com/sirupsen/logrus.(*Entry).write(0xc0001ef730)
	/.../go/src/github.com/mendersoftware/tenantadm/vendor/github.com/sirupsen/logrus/entry.go:294 +0x116
github.com/sirupsen/logrus.(*Entry).log(0xc000206230, 0x4, {0xc0004394d0, 0x22})
	/.../go/src/github.com/mendersoftware/tenantadm/vendor/github.com/sirupsen/logrus/entry.go:252 +0x445
github.com/sirupsen/logrus.(*Entry).Log(0xc000206230, 0x4, {0xc00051c7f8?, 0x1?, 0x1?})
	/.../go/src/github.com/mendersoftware/tenantadm/vendor/github.com/sirupsen/logrus/entry.go:304 +0x4f
github.com/sirupsen/logrus.(*Entry).Logf(0xc000206230, 0x4, {0x18ecb07?, 0x18d9822?}, {0xc00051c9d0?, 0x1f24dc0?, 0x1?})
	/.../go/src/github.com/mendersoftware/tenantadm/vendor/github.com/sirupsen/logrus/entry.go:349 +0x85
github.com/sirupsen/logrus.(*Entry).Infof(...)
	/.../go/src/github.com/mendersoftware/tenantadm/vendor/github.com/sirupsen/logrus/entry.go:362
github.com/mendersoftware/go-lib-micro/mongo/migrate.(*SimpleMigrator).Apply(0xc00051cce0, {0x1ab18d0, 0xc00012c008}, {0x0?, 0x0?, 0x0?}, {0xc00052e000?, 0x8, 0x8})
	/.../go/src/github.com/mendersoftware/tenantadm/vendor/github.com/mendersoftware/go-lib-micro/mongo/migrate/migrator_simple.go:124 +0xbf0
github.com/mendersoftware/tenantadm/store/mongo.(*DataStoreMongo).Migrate(0xc000524000, {0x18d0b46?, 0x0?}, {0x0?, 0x0?, 0x0?})
	/.../go/src/github.com/mendersoftware/tenantadm/store/mongo/datastore_mongo.go:631 +0x3ff
main.cmdMigrate(0x0?)
	/.../go/src/github.com/mendersoftware/tenantadm/main.go:320 +0x46b
main.runDaemon(0xc00051d0b7?)
	/.../go/src/github.com/mendersoftware/tenantadm/main.go:270 +0x2e
gopkg.in/urfave/cli%2ev1.HandleAction({0x17c1e20?, 0x192b7c8?}, 0xc0000166c0?)
	/.../go/src/github.com/mendersoftware/tenantadm/vendor/gopkg.in/urfave/cli.v1/app.go:490 +0x5a
gopkg.in/urfave/cli%2ev1.(*App).Run(0xc000503d40, {0xc0001181e0, 0x1, 0x1})
	/.../go/src/github.com/mendersoftware/tenantadm/vendor/gopkg.in/urfave/cli.v1/app.go:264 +0x725
main.doMain({0xc0001181e0, 0x1, 0x1})
	/.../go/src/github.com/mendersoftware/tenantadm/main.go:262 +0x1648
main.main()
	/.../go/src/github.com/mendersoftware/tenantadm/main.go:48 +0x2e
```

**Comment by alfrunes** [comment]
> 

Ouch! :see_no_evil: 
Sorry about the mess! I made the Configure interface a lot more fragile than I initially thought. I'll test it properly before I request another review.

**Comment by merlin-northern** [comment]
> > 
> 
> Ouch! 🙈 Sorry about the mess! I made the Configure interface a lot more fragile than I initially thought. I'll test it properly before I request another review.

do not worry, let me know how I can help.

**Comment by merlin-northern** [comment]
ok now I see that:
```bash
#mender_lite_run
INFO[0000] migration to version 1.1.0 skipped            db=tenantadm
INFO[0000] migration to version 1.2.0 skipped            db=tenantadm
INFO[0000] migration to version 1.3.0 skipped            db=tenantadm
INFO[0000] migration to version 1.4.0 skipped            db=tenantadm
INFO[0000] migration to version 1.4.1 skipped            db=tenantadm
INFO[0000] migration to version 1.4.2 skipped            db=tenantadm
INFO[0000] migration to version 1.4.3 skipped            db=tenantadm
INFO[0000] migration to version 1.4.4 skipped            db=tenantadm
INFO[0000] DB migrated to version 1.4.4                  db=tenantadm
INFO[0000] Tenant Administration Service starting up
INFO[0000] setting up prod middleware
INFO[0000] listening on :...
```

but it is different from master, which looks:
```bash
#mender_lite_run
INFO[2023-06-29T14:48:51+02:00] migration to version 1.1.0 skipped            db=tenantadm file=migrator_simple.go func="migrate.(*SimpleMigrator).Apply" line=124
INFO[2023-06-29T14:48:51+02:00] migration to version 1.2.0 skipped            db=tenantadm file=migrator_simple.go func="migrate.(*SimpleMigrator).Apply" line=124
INFO[2023-06-29T14:48:51+02:00] migration to version 1.3.0 skipped            db=tenantadm file=migrator_simple.go func="migrate.(*SimpleMigrator).Apply" line=124
INFO[2023-06-29T14:48:51+02:00] migration to version 1.4.0 skipped            db=tenantadm file=migrator_simple.go func="migrate.(*SimpleMigrator).Apply" line=124
INFO[2023-06-29T14:48:51+02:00] migration to version 1.4.1 skipped            db=tenantadm file=migrator_simple.go func="migrate.(*SimpleMigrator).Apply" line=124
INFO[2023-06-29T14:48:51+02:00] migration to version 1.4.2 skipped            db=tenantadm file=migrator_simple.go func="migrate.(*SimpleMigrator).Apply" line=124
INFO[2023-06-29T14:48:51+02:00] migration to version 1.4.3 skipped            db=tenantadm file=migrator_simple.go func="migrate.(*SimpleMigrator).Apply" line=124
INFO[2023-06-29T14:48:51+02:00] migration to version 1.4.4 skipped            db=tenantadm file=migrator_simple.go func="migrate.(*SimpleMigrator).Apply" line=124
INFO[2023-06-29T14:48:51+02:00] DB migrated to version 1.4.4                  db=tenantadm file=migrator_simple.go func="migrate.(*SimpleMigrator).Apply" line=139
INFO[2023-06-29T14:48:51+02:00] Tenant Administration Service starting up     file=main.go func=main.runDaemon line=277
INFO[2023-06-29T14:48:51+02:00] setting up prod middleware                    file=middleware.go func=main.SetupMiddleware line=107
INFO[2023-06-29T14:48:51+02:00] listening on :...
```

**Comment by alfrunes** [comment]
@merlin-northern Fixed the initialization issues and properly verified that all environment variables behave as expected. I slightly changed the caller format string to `<package>.<func>@<file>:<line>`.

**Comment by merlin-northern** [comment]
> @merlin-northern Fixed the initialization issues and properly verified that all environment variables behave as expected. I slightly changed the caller format string to `<package>.<func>@<file>:<line>`.

thanks!
```
INFO[2023-07-03T06:36:55+02:00] migration to version 1.1.0 skipped            caller="migrate.(*SimpleMigrator).Apply@migrator_simple.go:124" db=tenantadm
INFO[2023-07-03T06:36:55+02:00] migration to version 1.2.0 skipped            caller="migrate.(*SimpleMigrator).Apply@migrator_simple.go:124" db=tenantadm
INFO[2023-07-03T06:36:55+02:00] migration to version 1.3.0 skipped            caller="migrate.(*SimpleMigrator).Apply@migrator_simple.go:124" db=tenantadm
INFO[2023-07-03T06:36:55+02:00] migration to version 1.4.0 skipped            caller="migrate.(*SimpleMigrator).Apply@migrator_simple.go:124" db=tenantadm
INFO[2023-07-03T06:36:55+02:00] migration to version 1.4.1 skipped            caller="migrate.(*SimpleMigrator).Apply@migrator_simple.go:124" db=tenantadm
INFO[2023-07-03T06:36:55+02:00] migration to version 1.4.2 skipped            caller="migrate.(*SimpleMigrator).Apply@migrator_simple.go:124" db=tenantadm
INFO[2023-07-03T06:36:55+02:00] migration to version 1.4.3 skipped            caller="migrate.(*SimpleMigrator).Apply@migrator_simple.go:124" db=tenantadm
INFO[2023-07-03T06:36:55+02:00] migration to version 1.4.4 skipped            caller="migrate.(*SimpleMigrator).Apply@migrator_simple.go:124" db=tenantadm
INFO[2023-07-03T06:36:55+02:00] DB migrated to version 1.4.4                  caller="migrate.(*SimpleMigrator).Apply@migrator_simple.go:139" db=tenantadm
INFO[2023-07-03T06:36:55+02:00] Tenant Administration Service starting up     caller="main.runDaemon@main.go:277"
INFO[2023-07-03T06:36:55+02:00] setting up prod middleware                    caller="main.SetupMiddleware@middleware.go:107"
INFO[2023-07-03T06:36:55+02:00] listening on :58080
```
nice!

**Review by merlin-northern — APPROVED** [review]
lgtm. thanks for being patient. do we need to adjust some log parsers elsewhere to the new caller format? @oldgiova fyi: just so you know we have a new cool caller logs format.

**Pull request merged**